File: LeetCode - April/validParenthesisString.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        stack, star = [], []
        
        for index, char in enumerate(s):
            if char == '(':
                stack.append(index)
            elif char == "*":
                star.append(index)
            elif char == ')':
                if len(stack)>0:
                    stack.pop()
                elif len(star)>0:
                    star.pop()
                else:
                    return False
        
        while stack :
            if(star and (stack[-1] < star[-1]) ):
                star.pop()
            else:
                return False
            stack.pop()
        return len(stack)==0

File: LeetCode - April/validParenthesisString.java
class Solution {
    public boolean checkValidString(String s) {
        
        Stack<Integer> stack = new Stack<Integer>();
        Stack<Integer> star = new Stack<Integer>(); 
    
        for(int i = 0; i < s.length(); i++){
            
            if(s.charAt(i) == '*')
                star.push(i);
            
            else if(s.charAt(i) == '(')
                stack.push(i);
            
            else{
                if(!stack.isEmpty())
                    stack.pop();
                else if(!star.isEmpty())
                    star.pop();
                else
                    return false;
            }
            
        }
        
        while(!stack.isEmpty()){
            if(!star.isEmpty() && star.peek() > stack.peek())
                star.pop();
            else
                return false;
            stack.pop();
        }
        
        return true;
    }
}